src/exec/wait_for_children.c

```diff
@@ -15,22 +15,44 @@
 
 #include "ctx.h"
 
-void	wait_for_children(pid_t last_pid, t_ctx *ctx)
+static void	update_exit_status(pid_t wpid, int status,
+				pid_t last_pid, t_ctx *ctx)
+{
+	if (wpid == last_pid)
+	{
+		if (WIFEXITED(status))
+			ctx->exit_status = WEXITSTATUS(status);
+		else if (WIFSIGNALED(status))
+			ctx->exit_status = WTERMSIG(status) + 128;
+	}
+}
+
+static bool	wait_for_single_child(pid_t last_pid, t_ctx *ctx)
 {
 	int		status;
 	pid_t	wpid;
 
-	while (true)
+	wpid = waitpid(-1, &status, 0);
+	if (wpid == -1)
 	{
-		wpid = waitpid(-1, &status, 0);
-		if (wpid == -1)
-			return ;
-		if (wpid == last_pid)
-		{
-			if (WIFEXITED(status))
-				ctx->exit_status = WEXITSTATUS(status);
-			else if (WIFSIGNALED(status))
-				ctx->exit_status = WTERMSIG(status) + 128;
-		}
+		if (errno == ECHILD)
+			return (false);
+		print_error("waitpid", strerror(errno));
+		return (false);
 	}
+	if (last_pid != -1)
```

Comment:
suggestion to simplify `update_exit_status()` to the following prototype:

`static void	update_exit_status(int status, t_ctx *ctx)`

```suggestion
	if (last_pid != -1  && wpid == last_pid)
```